rfc/system/5818-decentralized-agent-config-mgmt.md

```diff
@@ -0,0 +1,147 @@
+- **Separation of concerns**: Teams can *fully* independently manage their own custom agent configurations.
+- **Decentralized management**: Custom agents can be deployed and managed independently from the core Flyte infrastructure, yet operate as a cohesive unit.
+- **Flexibility**: FlytePropeller can dynamically adapt to varying configurations without requiring centralized maintenance of all settings.
+
+## 2. Motivation
+
+In enterprise environments, the team managing the core Flyte infrastructure is usually different from the teams developing custom agents/plugins. These teams operate with separate development cycles and CI/CD pipelines. For instance, a team writing its own custom agent would create its own service account, Docker image, and deploy it in a separate namespace with its own monitoring dashboard.
+
+However, the final step of deployment—updating the Propeller configuration—remains centralized. This leads to issues like:
+1. **Out-of-sync deployment**: When a custom agent endpoint needs to be updated, it requires coordination between teams for the deployment of both the agent and the Propeller configuration update, leading to potential mismatches and outage.
+2. **Coordination overhead**: Misaligned rollouts could result in the Propeller being unable to reach the agent, potentially causing blast radius of impact due to propeller is down.
+
+To address these challenges and fully enable the decentralized custom agent framework, this RFC proposes a decentralized approach to custom agent configuration management, allowing:
+- **Separation of concerns**: Teams can manage their own configurations independently.
+- **Decentralized management**: Custom agents can be deployed and managed separately from Flyte’s core infrastructure.
+- **Flexibility**: FlytePropeller can adapt to various configurations dynamically.
+
+## 3. Proposed Implementation
+
+Currently, Flyte [parses configuration files under `/etc/flyte/config` and decodes them](https://github.com/flyteorg/flyte/tree/master/flytestdlib/config/viper) for use later. In decentralized management, the custom agent configuration might not be accessible in the same volume or mount, meaning traditional parsing from the Propeller pod's filesystem won't be possible. 
+
+Instead, the configuration could be managed dynamically via Kubernetes APIs, allowing FlytePropeller to discover, validate, and merge custom agent configurations at runtime.
+
+### Discovery of Custom Agent ConfigMaps
+Implement a mechanism to systematically discover and collect custom agent ConfigMaps across namespaces or predefined locations. This ensures any new custom agent configurations are properly detected and processed.
+
+For example, it is not unreasonable to ask the custom agent team to create configmap named with specific pattern, such as `flyteplugins-agentservice-` so that the propeller can list the config maps in the cluster via K8s API.
+
```

Comment:
I think the canonic way in Kubernetes to achieve this would be to create a "watch" for a specific label selector.
I'd prefer this over a naming convention.

Reply:
cc @shuyingliang